### PR TITLE
auto generate tests from samples

### PR DESCRIPTION
**Pull request by ScottSuarez**
**Release Note Template for Downstream PRs (will be copied)**

```release-note:none
```

### CODE REVIEW

File: tpgtools/README.md [hunk 1]
```diff
@@ -6,7 +6,7 @@ to describe the available types, and `tpgtools` uses those to construct
 Terraform resource schemas.
 
 ## Usage
```
Comment (c2thorn) [inline tpgtools/README.md:8]: We should probably add some instructions on how to use the samples folder here, if you want to do that in a follow up or something

File: tpgtools/documentation.go [hunk 2]
```diff
@@ -86,8 +86,12 @@ func mergeProperties(ga, beta []Property) []Property {
 }
 
 func generateResourceWebsiteFile(res *Resource, resources map[Version][]*Resource, version *Version) {
-
 	res = mergeResource(res, resources, version)
+
+	if len(res.DocSamples()) <= 0 {
```
Comment (c2thorn) [inline tpgtools/documentation.go:91]: This is probably a good way to determine when we should use MMV1 or tpgtools for documentation right now, but could confuse someone who doesn't know about it. can we log that tpgtools is skipping doc generation and why here?

File: tpgtools/main.go [hunk 3]
```diff
@@ -35,6 +35,7 @@ var fPath = flag.String("path", "", "path to the root service directory holding
 var tPath = flag.String("overrides", "", "path to the root directory holding overrides files")
 var cPath = flag.String("handwritten", "handwritten", "path to the root directory holding handwritten files to copy")
 var oPath = flag.String("output", "", "path to output generated files to")
+var sPath = flag.String("samples", "samples", "path to the directory holding the samples and substitution data")
```
Comment (c2thorn) [inline tpgtools/main.go:38]: non-related:
This is consistent with `fpath` and `tpath` but I wonder why we offer the ability to override the paths. Do we really expect anyone to give a non-fixed path?

File: tpgtools/main.go [hunk 4]
```diff
@@ -384,6 +416,49 @@ func generateSweeperFile(res *Resource) {
...
+		glog.Exit(err)
+	}
+
+	if err != nil {
+		glog.Exit(err)
+	}
+
+	formatted, err := formatSource(&contents)
+	if err != nil {
+		err := ioutil.WriteFile("./temp/temp.go", []byte(contents.String()), 0644)
```
Comment (c2thorn) [inline tpgtools/main.go:447]: Doesn't this overwrite the original format error?

File: tpgtools/resource.go [hunk 5]
```diff
@@ -508,3 +545,205 @@ func readVirtualField(vf VirtualFieldDetails) Property {
...
+	samples := []Sample{}
+	handWritten := r.loadHandWrittenSamples()
+	dclSamples := r.loadDCLSamples()
+	samples = append(samples, dclSamples...)
+	samples = append(samples, handWritten...)
+	return samples
+}
+
+func (r *Resource) loadHandWrittenSamples() []Sample {
+	sampleFriendlyFolder := r.getSampleAccessoryFolder()
```
Comment (c2thorn) [inline tpgtools/resource.go:600]: nit: worth being consistent with either "Friendly" or "Accessory"?

File: tpgtools/sample.go [hunk 6]
```diff
@@ -0,0 +1,377 @@
...
+		resourceName = strings.Title(fileParts[1])
+	} else {
+		return nil, fmt.Errorf("Invalid sample dependency file name: %s", fileName)
+	}
+
+	if localname == "" {
+		localname = fileParts[0]
+	}
+	dclResourceType := product + resourceName
+	// terraformResourceType := fmt.Sprintf("google_%s_%s", fileParts[1], fileParts[2])
```
Comment (c2thorn) [inline tpgtools/sample.go:117]: Think I left this in
```suggestion
```

File: tpgtools/sample.go [hunk 7]
```diff
@@ -0,0 +1,377 @@
...
+	for index := range s.DependencyList {
+		err := s.ReplaceReferences(&s.DependencyList[index])
+		if err != nil {
+			return "", fmt.Errorf("Could not generate HCL template for %s: %s", *s.Name, err)
+		}
+		// Skip appending the primary resource, it should go last
+		// if strings.Contains(*s.PrimaryResource, s.DependencyList[index].FileName) {
+		// 	primaryIndex = index
+		// 	continue
+		// }
```
Comment (c2thorn) [inline tpgtools/sample.go:211]: more stuff I left in
```suggestion
```

File: tpgtools/sample.go [hunk 8]
```diff
@@ -0,0 +1,377 @@
...
+	}
+	return nil
+}
+
+func (s Sample) generateHCLTemplate() (string, error) {
+	if len(s.DependencyList) == 0 {
+		return "", fmt.Errorf("Could not generate HCL template for %s: there are no dependencies", *s.Name)
+	}
+	var hcl string
+	// var primaryIndex int
```
Comment (c2thorn) [inline tpgtools/sample.go:201]:
```suggestion
```

File: tpgtools/sample.go [hunk 9]
```diff
@@ -0,0 +1,377 @@
...
+		}
+		// Skip appending the primary resource, it should go last
+		// if strings.Contains(*s.PrimaryResource, s.DependencyList[index].FileName) {
+		// 	primaryIndex = index
+		// 	continue
+		// }
+		hcl = fmt.Sprintf("%s%s\n", hcl, s.DependencyList[index].HCLBlock)
+	}
+
+	// hcl = fmt.Sprintf("%s%s", hcl, s.DependencyList[primaryIndex].HCLBlock)
```
Comment (c2thorn) [inline tpgtools/sample.go:215]:
```suggestion
```

File: tpgtools/samples/assuredworkloads/workload/workload_post_destory.go [hunk 10]
```diff
@@ -0,0 +1,78 @@
+// deleteAssuredWorkloadProvisionedResources deletes the resources provisioned by
```
Comment (c2thorn) [inline tpgtools/samples/assuredworkloads/workload/workload_post_destory.go:1]: filename should be `workload_post_destroy`?

File: tpgtools/samples/assuredworkloads/workload/workload_post_destory.go [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+	return func(s *terraform.State) error {
+		config := googleProviderConfig(t)
+		timeout := *schema.DefaultTimeout(4 * time.Minute)
+		for _, rs := range s.RootModule().Resources {
+			if rs.Type != "google_assured_workloads_workload" {
+				continue
+			}
+			resourceAttributes := rs.Primary.Attributes
+			n, err := strconv.Atoi(resourceAttributes["resources.#"])
+			log.Printf("[DEBUG]found %v resources\n", n)
```
Comment (c2thorn) [inline tpgtools/samples/assuredworkloads/workload/workload_post_destory.go:13]: Do we want these log statements for the "happy" path?

**Review by c2thorn — CHANGES_REQUESTED** [review]
minor stuff

File: tpgtools/README.md (hunk 1 above)
Comment (ScottSuarez) [reply tpgtools/README.md:8]: Going to do that in a follow up since things are changing soon

File: tpgtools/main.go (hunk 3 above)
Comment (ScottSuarez) [reply tpgtools/main.go:38]: going to be removing this anyways in the future when I migrate to samples.. leaving for now

File: tpgtools/main.go (hunk 4 above)
Comment (ScottSuarez) [reply tpgtools/main.go:447]: yeah this was debugging code... I've removed it

**Review by c2thorn — APPROVED** [review]

File: tpgtools/resource.go (hunk 5 above)
Comment (ScottSuarez) [reply tpgtools/resource.go:600]: accessory seems more reliable

File: tpgtools/samples/assuredworkloads/workload/workload_post_destory.go (hunk 10 above)
Comment (ScottSuarez) [reply tpgtools/samples/assuredworkloads/workload/workload_post_destory.go:1]: changed

File: tpgtools/samples/assuredworkloads/workload/workload_post_destory.go (hunk 11 above)
Comment (ScottSuarez) [reply tpgtools/samples/assuredworkloads/workload/workload_post_destory.go:13]: I think this is relevant. It's debug logs which already contain the rest responses. I think this makes the data more readable :)

File: tpgtools/documentation.go (hunk 2 above)
Comment (ScottSuarez) [reply tpgtools/documentation.go:91]: done

**Comment by rileykarson** [comment]
I think this broke TPG @ScottSuarez

**Pull request merged**